### ISSUE TITLE
[Question] Require multiple roles in a policy

### ISSUE THREAD

**Issue by derberni** [casbin/casbin#1451]
**What's your scenario? What do you want to achieve?**
We want to use a RBAC model, however in order to access an object, the subject must be associated with more than one role. As an example, we want to limit access to an object to users that have the `role:employee` and `role:feature_1`.

A naive approach to model this as a policy would be something like:
`p, "role:employee,role:feature_1", /file, read`
However, that obviously does not work with the `g(_, _)` function to check if the subject matches the role.

Is there currently any way to do this with casbin? Could you in theory replace `g(_, _)` by a custom implementation that internally uses `g(_, _)` but applies it to all roles listet in the "role:A,role:B,role:C" string?

**Comment by onnttf** [comment]
Why not consider splitting these two roles into two p?

```plaintext
p, "role:employee", /file, read
p, "role:feature_1", /file, read
```

Or set it into one p, and then set the inheritance relationship between roles?

```plaintext
p, "role:employee", /file, read
g, "role:feature_1", "role:employee"
```

**Comment by derberni** [comment]
That would only require the subject to have one of the two roles. But we require them to have both roles.

**Comment by onnttf** [comment]
> That would only require the subject to have one of the two roles. But we require them to have both roles.

Like this below?

```plaintext
p, role:employee, /file, read
p, role:feature_1, /file, read
g, user, role:employee
g, user, role:feature_1
```

**Comment by derberni** [comment]
Yes, policy wise that is what I would want to define. However with the limited policy effects casbin supports, for example "allow-override", once one of the policies matches access is granted. For these policies to work, we would need a "all-allow" policy effect

**Comment by onnttf** [comment]
> Yes, policy wise that is what I would want to define. However with the limited policy effects casbin supports, for example "allow-override", once one of the policies matches access is granted. For these policies to work, we would need a "all-allow" policy effect

Sorry, I don't understand your requirements.
Once one meets the rules, then all will be allowed to pass. This is normal.
Do you want both strategies to hit before you allow them to pass?

Or try to use custom function to handle it?

[Functions](https://casbin.org/docs/function)

**Comment by hsluoyz** [comment]
@derberni You can use the "ABAC with policy rule" model: https://casbin.org/docs/abac#scaling-the-model-for-complex-and-large-numbers-of-abac-rules

and use this policy:

```p, g(r.sub, "role:employee") && g(r.sub, "role:feature_1"), /file, read```
